x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinTypes.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.plan.logical.join;
+
+import org.elasticsearch.xpack.ql.expression.Attribute;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Utility class defining the concrete types of joins supported by ESQL.
+ */
+public class JoinTypes {
+
+    private JoinTypes() {}
+
+    public static CoreJoinType INNER = CoreJoinType.INNER;
+    public static CoreJoinType LEFT = CoreJoinType.LEFT;
+    public static CoreJoinType RIGHT = CoreJoinType.RIGHT;
+    public static CoreJoinType FULL = CoreJoinType.FULL;
+    public static CoreJoinType CROSS = CoreJoinType.CROSS;
+
+    /**
+     * The predefined core join types. Implements as enum for easy comparison and serialization.
+     */
+    public enum CoreJoinType implements JoinType {
+        INNER("INNER"),
+        LEFT("LEFT OUTER"),
+        RIGHT("RIGHT OUTER"),
+        FULL("FULL OUTER"),
+        CROSS("CROSS");
+
+        private final String name;
+
+        CoreJoinType(String name) {
+            this.name = name;
+        }
+
+        @Override
+        public String joinName() {
+            return name;
+        }
+    }
+
+    /**
+     * Join type for the USING clause - shorthand for defining an equi-join (equality join meaning the condition checks if columns across
+     * each side of the join are equal).
+     * One important difference is that the USING clause does not include the join columns in the result set.
+     */
+    public static class UsingJoinType implements JoinType {
```

Comment:
It does feel weird to have USING be something different from a join. Like, maybe it's just

```
record JoinConfig(List<Attribute> mergedColumns, List<Expression> extraConditions, JoinType left/right/outer/whatever)
```

And right now we don't support any `extraConditions`. And `mergedColumns` has to be non-empty unless we're doing the FULL join on a single row thing.